app/scripts/background.js

```diff
@@ -981,45 +981,88 @@ export function setupController(
           connectionStream: portStreamForCookieHandlerPage,
         });
       }
-      connectExternalExtension(remotePort);
+
+      const portStream =
+        overrides?.getPortStream?.(remotePort) || new PortStream(remotePort);
+
+      connectEip1193(portStream, remotePort.sender);
+
+      if (process.env.MULTICHAIN_API && isFirefox) {
+        const mux = setupMultiplex(portStream);
+        mux.ignoreStream(METAMASK_EIP_1193_PROVIDER);
+
+        connectCaipMultichain(
+          mux.createStream(
+            METAMASK_CAIP_MULTICHAIN_PROVIDER,
+            remotePort.sender,
+          ),
+        );
+      }
     }
   };
 
-  // communication with page or other extension
-  connectExternalExtension = (remotePort) => {
+  /**
+   * Connects a externally_connecatable Port to the MetaMask controller.
+   * This method identifies dapp clients and connects them differently from extension clients.
+   *
+   * @param {Port} remotePort - The port provided by a new context.
+   */
+  connectExternallyConnectable = (remotePort) => {
     const portStream =
       overrides?.getPortStream?.(remotePort) || new PortStream(remotePort);
+
+    const isDappConnecting = remotePort.sender.tab?.id;
+    if (isDappConnecting && process.env.MULTICHAIN_API) {
+      if (metamaskBlockedPorts.includes(remotePort.name)) {
+        return;
+      }
+
+      // this is triggered when a new tab is opened, or origin(url) is changed
+      if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
```

Comment:
We already know that `remotePort.sender` and `remotePort.sender.tab` is defined, because the `isDappConnecting` connecting check requires it. If `remotePort.sender` is `undefined`, the `isDappConnecting` will throw an `Uncaught TypeError: Cannot read properties of undefined (reading 'tab')` error.

Note: the chrome.runtime.Port types document it as `Optional. This property will only be present on ports passed to onConnect/onConnectExternal listeners.`. I haven't dug into if this is used outside those listeners or not.

Reply:
I've removed the guard here https://github.com/MetaMask/metamask-extension/pull/30142/commits/d907c7e4250b19996135ac1e95134845a4432155

I wasn't clear if you were suggesting that an optional chaining operator be added to remotePort.sender since it should be always be defined as we use it

Reply:
whatever make sense is fine. changes look good!